### PR TITLE
tests: add spread test for assertions in prepare-image's extra-snaps

### PR DESCRIPTION
**Pull request by fgimenez**
Hi Michael, this is the test I mentioned, it can be executed with: `spread -v -debug qemu:ubuntu-16.04-64:tests/nested/extra-snaps-assertions`. The test downloads core, creates a new image with it as `--extra-snaps`, boots it with qemu (doing all the required magic for putting in place a system-user assertion) and checks that the core in the image has a proper revision number (not x1). The last part, the checks on the created image, is performed in `tests/nested/prepare-image-check`, maybe we could add more checks there.

All the nested suite is marked as manual, so that this test won't be included in regular PR executions, we should setup a spread-cron branch for having them executed as needed (either in response to events like merges to master or cron-based).

### CODE REVIEW

File: tests/nested/extra-snaps-assertions/task.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    mkdir -p /tmp/work-dir
+    snap install --devmode --beta ubuntu-image
+
+    snap download core
+
+    /snap/bin/ubuntu-image --image-size 3G $TESTSLIB/assertions/nested-${NESTED_ARCH}.model --channel $CORE_CHANNEL --output ubuntu-core.img --extra-snaps core_*.snap
+    mv ubuntu-core.img /tmp/work-dir
+
+    # create assertion block device
+    genisoimage -volid cidata -joliet -rock -o assertions.disk $TESTSLIB/assertions/auto-import.assert
```
Comment (mvo5) [inline tests/nested/extra-snaps-assertions/task.yaml:36]: Do we need this auto-import.assert file here?

File: tests/nested/extra-snaps-assertions/task.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        ;;
+    *)
+        echo "unsupported architecture"
+        exit 1
+        ;;
+    esac
+
+    # create ubuntu-core image
+    mkdir -p /tmp/work-dir
+    snap install --devmode --beta ubuntu-image
```
Comment (mvo5) [inline tests/nested/extra-snaps-assertions/task.yaml:26]: Installing this ubuntu-image snap will bundle its own /usr/bin/snap - so this test (as is) will not work, will it? We can probably tweak it by `mount -o bind /usr/bin/snap /snap/ubuntu-image/current/bin/snap` ?

File: tests/nested/extra-snaps-assertions/task.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+restore: |
+    snap remove ubuntu-image
+    . $TESTSLIB/systemd.sh
+    systemd_stop_and_destroy_unit nested-vm
+    rm -rf /tmp/work-dir
+
+execute: |
+    execute_remote(){
+        sshpass -p ubuntu ssh -p 8022 -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no user1@localhost "$*"
```
Comment (mvo5) [inline tests/nested/extra-snaps-assertions/task.yaml:49]: nice

File: tests/nested/extra-snaps-assertions/task.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    execute_remote "echo test:ubuntu | sudo chpasswd"
+    execute_remote "echo 'test ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/test-user"
+
+    cd $SPREAD_PATH
+
+    curl -s -O https://niemeyer.s3.amazonaws.com/spread-amd64.tar.gz && tar xzvf spread-amd64.tar.gz && rm -f spread-amd64.tar.gz && mv spread /tmp/work-dir
+
+    set +x
+    export SPREAD_EXTERNAL_ADDRESS=localhost:8022
+    rm -f snapd-state.tar.gz npm-debug.log
```
Comment (mvo5) [inline tests/nested/extra-snaps-assertions/task.yaml:74]: Where does npm come from?

File: tests/nested/prepare-image-check/task.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+manual: true
+execute: |
+    echo "Wait for first boot to be done"
+    while ! snap changes | grep -q "Done.*Initialize system state"; do sleep 1; done
+
+    echo "We have a model assertion"
+    snap known model|grep "series: 16"
+
+    echo "Make sure core has an actual revision"
+    snap list | MATCH "^core +[0-9]+\.[0-9]+ +[0-9]+ +canonical +\-"
```
Comment (mvo5) [inline tests/nested/prepare-image-check/task.yaml]: One extra test that would be nice is to check if `snap info |MATCH "tracking: (stable|candidate|beta|edge)"` (not sure if MATCH supports this syntax. I.e. ensure we are tracking a channel.

**Review by mvo5 — COMMENTED** [review]
Thanks for this branch, the nested suite is real magic, nice work! I added some comments, but this looks very good.

File: tests/nested/extra-snaps-assertions/task.yaml (hunk 2 above)
Comment (fgimenez) [reply tests/nested/extra-snaps-assertions/task.yaml:26]: Brilliant! I was using an ubuntu-image built using your branch as source for snapd, bind mounting works like a charm :) changes pushed, the test is passing now, you can try it with `spread -v qemu:ubuntu-16.04-64:tests/nested/extra-snaps-assertions`

File: tests/nested/extra-snaps-assertions/task.yaml (hunk 1 above)
Comment (fgimenez) [reply tests/nested/extra-snaps-assertions/task.yaml:36]: AFAIK yes, if not the nested instance is not accessible at all, we would need to go through the console-conf initialization, this system-user assertion allows us to access as user1:ubuntu, which is used later to configure the test user required by spread.

File: tests/nested/extra-snaps-assertions/task.yaml (hunk 4 above)
Comment (fgimenez) [reply tests/nested/extra-snaps-assertions/task.yaml:74]: No idea, the first runs were failing because the state tarball was being left at the root of the project, while looking there I saw that npm file there too (~800kb), I'll dig deeper to find the cause.

**Comment by fgimenez** [comment]
@mvo5 thanks! comments addressed, pls take another look

File: tests/nested/extra-snaps-assertions/task.yaml (hunk 1 above)
Comment (mvo5) [reply tests/nested/extra-snaps-assertions/task.yaml:36]: Aha, nice. Thank you
